libnetwork/ipamapi/contract.go

```diff
@@ -63,6 +60,41 @@ type Ipam interface {
 	IsBuiltIn() bool
 }
 
+type PoolRequest struct {
+	// AddressSpace is a mandatory field which denotes which block of pools
+	// should be used to make the allocation. This value is opaque, and only
+	// the IPAM driver can interpret it. Each driver might support a different
+	// set of AddressSpace.
+	AddressSpace string
+	// Pool is a prefix in CIDR notation. It's non-mandatory. When specified
+	// the Pool will be statically allocated. The Pool is used for dynamic
+	// address allocation -- except when SubPool is specified.
+	Pool string
+	// SubPool is a subnet from Pool, in CIDR notation too. It's non-mandatory.
+	// When specified, it represents the subnet where addresses will be
+	// dynamically allocated. It can't be specified if Pool isn't specified.
+	SubPool string
+	// Options is a map of opaque k/v passed to the driver. It's non-mandatory.
+	// Drivers are free to ignore it.
+	Options map[string]string
+	// V6 indicates which address family should be used to dynamically allocate
+	// a prefix (ie. when Pool isn't specified).
+	V6 bool
+}
+
+type AllocatedPool struct {
+	// PoolID represents the ID of the allocated pool. Its value is opaque and
+	// shouldn't be interpreted by anything but the IPAM driver that generated
+	// it.
+	PoolID string
+	// Pool is the allocated prefix.
+	Pool netip.Prefix
+	// Meta represents a list of extra IP addresses automatically reserved
+	// during the pool allocation. These are generally keyed by well-known
+	// strings defined in the netlabel package.
+	Meta map[string]string
```

Comment:
If the values are always IP addresses, why are they stringly-typed? Or is the doc comment misleadingly describing how the default IPAM driver happens to populate a generic bag of data with implementation-defined semantics?

Reply:
Yeah, I'll update this comment -- it's misleading.

I'm also wondering how useful it is. No builtin driver set any `Meta`, only remote drivers might do that. And it's used only in two places:

https://github.com/moby/moby/blob/16b2c223ce1c2381b6be67aaa218f3b02746e01c/libnetwork/cnmallocator/networkallocator.go#L905-L910

https://github.com/moby/moby/blob/16b2c223ce1c2381b6be67aaa218f3b02746e01c/libnetwork/network.go#L1619-L1623

So, I think it'd even be safe to replace it with an optional `Gateway`.